### PR TITLE
Feature/single export

### PR DESCRIPTION
**Pull request by guy-har**
Draft,
Contains todos
Need to add ddl to migration and change to work with one DB instance.

### CODE REVIEW

File: export/export.go [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+func getExportID(repo, branch, commitRef string) (string, error) {
+	nid, err := nanoid.Nanoid()
+	if err != nil {
+		return "", err
+	}
+	return fmt.Sprintf("%s-%s-%s-%s", repo, branch, commitRef, nid), nil
+}
+
+// ExportBranchStart inserts a start task on branch and changes branch export state to pending
+// if export on  already in progress will return error
```
Comment (arielshaqed) [inline export/export.go:23]: Could you rephrase this comment please?  I found it a bit hard to follow -- maybe just adding punctuation will make it easier for me!

File: export/export_handler.go [hunk 2]
```diff
@@ -41,6 +49,121 @@ func PathToPointer(path string) (block.ObjectPointer, error) {
...
+		taskData, err := tasksGenerator.Add(diffs)
+		if err != nil {
+			return err
+		}
+		// add taskData tasks
+		err = h.parade.InsertTasks(context.Background(), taskData)
+		if err != nil {
+			return err
+		}
+		if hasMore && len(diffs) > 0 {
```
Comment (arielshaqed) [inline export/export_handler.go:103]: Can we have `hasMore && len(diffs) == 0`?!  (Note that in that case it looks like an infinite loop, so not sure what this extra caution does here.  If it's a worry, it should be an error message -- maybe even fatal.)

File: export/tasks_generator.go [hunk 3]
```diff
@@ -105,12 +122,19 @@ func (exportID taskIDGenerator) finishedTaskID() parade.TaskID {
 	return parade.TaskID(fmt.Sprintf("%s:finished", exportID))
 }
 
-func (exportID taskIDGenerator) copyTaskID(physicalAddress string) parade.TaskID {
-	return parade.TaskID(fmt.Sprintf("%s:copy:%s", exportID, physicalAddress))
+func (exportID taskIDGenerator) startedTaskID() parade.TaskID {
+	return parade.TaskID(fmt.Sprintf("%s:started", exportID))
```
Comment (arielshaqed) [inline export/tasks_generator.go:126]: This is a simple past-tense verb, whereas in other task IDs we use an imperative verb.  Could you do that here, too?  (Not sure which imperative form is intended here!)

File: export/tasks_generator.go [hunk 4]
```diff
@@ -236,32 +260,59 @@ type TasksGenerator struct {
...
+		idGen:                 taskIDGenerator(exportID),
+		successTasksGenerator: NewSuccessTasksTreeGenerator(exportID, generateSuccessFor, makeDestination, finishBody),
 	}
 }
 
 // Add translates diffs into many tasks and remembers "generate success" tasks for Finish.  It
 // returns some tasks that can already be added.
 func (e *TasksGenerator) Add(diffs catalog.Differences) ([]parade.TaskData, error) {
-	const initialSize = 1_000
+	const initialSize = 1000
```
Comment (arielshaqed) [inline export/tasks_generator.go:315]: Why?  I _really_ like this form, it is much easier for me to read large numbers when their digits are grouped in 3s.

File: export/tasks_generator_test.go [hunk 5]
```diff
@@ -327,59 +335,64 @@ func TestTasksGenerator_Simple(t *testing.T) {
 	}
 }
 
+func getGeneratedID(path string) string {
```
Comment (arielshaqed) [inline export/tasks_generator_test.go:338]: This copies code from the system under test.  Consider just exporting it -- maybe inside a `ForTesting` object.  Or strip the generated "random" values before comparing.

File: parade/ddl.sql [hunk 6]
```diff
@@ -8,7 +8,7 @@ CREATE TYPE task_status_code_value AS ENUM (
 );
 
 CREATE TABLE IF NOT EXISTS tasks (
-    id VARCHAR(64) NOT NULL PRIMARY KEY, -- nanoid
+    id VARCHAR(128) NOT NULL PRIMARY KEY, -- nanoid
```
Comment (arielshaqed) [inline parade/ddl.sql:11]: Note that this "128"\ appears in numerous other places.  And let's get rid of `nanoid` in the comment -- it never was.

Thinking about it, we should probably get rid of the length limitation.  Per https://www.postgresql.org/docs/12/datatype-character.html there is actually no advantage to this limitation, it only saves space when it truncates :scream: .

**Review by arielshaqed — COMMENTED** [review]
Cool stuff!

I think it's really clean good code.

File: export/tasks_generator.go (hunk 4 above)
Comment (guy-har) [reply export/tasks_generator.go:318]: GoLands linter threw an exception (it isn't updated to work with this form yet), which caused the auto formatter to stop working. so because it was a small number I allowed myself to change it.
I could change it back

File: api/api_controller.go [hunk 7]
```diff
@@ -2222,6 +2228,29 @@ func (c *Controller) ExportGetContinuousExportHandler() exportop.GetContinuousEx
 	})
 }
 
+func (c *Controller) ExportExecuteContinuousExport() exportop.ExecuteContinuousExportHandler {
+	return exportop.ExecuteContinuousExportHandlerFunc(func(params exportop.ExecuteContinuousExportParams, user *models.User) middleware.Responder {
+		deps, err := c.setupRequest(user, params.HTTPRequest, []permissions.Permission{
+			{
+				Action:   permissions.CreateBranchAction,
```
[the remark this answers is not in the document]
Comment (guy-har) [reply api/api_controller.go:2235]: Agree, changing it. 
Thanks

File: catalog/cataloger_export.go [hunk 8]
```diff
@@ -136,6 +136,23 @@ func (c *cataloger) PutExportConfiguration(repository string, branch string, con
...
 var ErrExportFailed = errors.New("export failed")
 
+func (c *cataloger) ExportState(repo, branch, newRef string, cb func(oldRef string, state CatalogBranchExportStatus) (newState CatalogBranchExportStatus, newMessage *string, err error)) error {
+	_, err := c.db.Transact(func(tx db.Tx) (interface{}, error) {
+		oldRef, state, err := c.ExportStateMarkStart(tx, repo, branch, newRef)
+		if err != nil {
+			return nil, err
+		}
+
+		newState, newMsg, err := cb(oldRef, state)
```
[the remark this answers is not in the document]
Comment (guy-har) [reply catalog/cataloger_export.go:146]: In case that `newState != ExportStatusFailed` 
The DB will return an error in case that `msg != nil`

File: export/export_handler.go (hunk 2 above)
Comment (guy-har) [reply export/export_handler.go]: I missed that,
Thanks

File: export/export.go (hunk 1 above)
Comment (guy-har) [reply export/export.go]: Done

File: export/tasks_generator_test.go (hunk 5 above)
Comment (guy-har) [reply export/tasks_generator_test.go:338]: 👍

File: api/api_controller.go [hunk 9]
```diff
@@ -2222,6 +2228,29 @@ func (c *Controller) ExportGetContinuousExportHandler() exportop.GetContinuousEx
...
+				WithPayload(responseErrorFrom(err))
+		}
+		deps.LogAction("execute_continuous_export")
+
+		err = export.ExportBranchStart(deps.Parade, deps.Cataloger, params.Repository, params.Branch)
+		if err != nil {
+			return exportop.NewExecuteContinuousExportDefault(http.StatusInternalServerError).
+				WithPayload(responseErrorFrom(err))
+		}
+		return exportop.NewExecuteContinuousExportCreated()
```
[the remark this answers is not in the document]
Comment (guy-har) [reply api/api_controller.go]: sounds good, I added it.

File: api/api_controller.go [hunk 10]
```diff
@@ -191,7 +197,7 @@ func (c *Controller) Configure(api *operations.LakefsAPI) {
 
 	api.ExportGetContinuousExportHandler = c.ExportGetContinuousExportHandler()
 	api.ExportSetContinuousExportHandler = c.ExportSetContinuousExportHandler()
-
+	api.ExportExecuteContinuousExportHandler = c.ExportExecuteContinuousExport()
```
[the remark this answers is not in the document]
Comment (guy-har) [reply api/api_controller.go]: 👌

File: export/tasks_generator.go (hunk 3 above)
Comment (guy-har) [reply export/tasks_generator.go]: 👌

File: parade/ddl.sql (hunk 6 above)
Comment (guy-har) [reply parade/ddl.sql]: Done, I removed the length limitation

**Comment by guy-har** [comment]
@arielshaqed PTAL

File: export/export_handler_test.go [hunk 11]
```diff
@@ -150,3 +151,45 @@ func TestTouch(t *testing.T) {
...
+	}{
+		{
+			name:                   "one regex",
+			lastKeysInPrefixRegexp: []string{".*\\.success$"},
+			expectTrue:             []string{"a.success", "other.success"},
+			expectFalse:            []string{"dfd", "a.suc", "a.successer"},
+			want:                   nil,
+		},
+		{
+			name:                   "two regexes",
```
Comment (arielshaqed) [inline export/export_handler_test.go:172]:
```suggestion
			name:                   "two regexps",
```
(one advantage of "regexp" is that you don't have smart alecs suggesting "regexen" as the plural.

**Review by arielshaqed — APPROVED** [review]
Thanks!

I think we're there, right?
